pkg/pgmodel/ingestor/dispatcher.go

```diff
@@ -178,6 +189,250 @@ func (p *pgxDispatcher) Close() {
+			event.Timestamp().AsTime(),
+			getUUID(traceID),
+			spanIDInt,
+			event.Name(),
+			i,
+			string(jsonTags),
+			event.DroppedAttributesCount(),
+		)
+
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func convertByteArrayToInt64(buf [8]byte) int64 {
+	ux := binary.BigEndian.Uint64(buf[:])
+
+	x := int64(ux >> 1)
+	if ux&1 != 0 {
+		x = ^x
+	}
+
+	return x
+}
+
+func getUUID(buf [16]byte) pgtype.UUID {
+	return pgtype.UUID{
+		Bytes:  buf,
+		Status: pgtype.Present,
+	}
+}
+
+func getEventTimeRange(events pdata.SpanEventSlice) (result pgtype.Tstzrange) {
+	if events.Len() == 0 {
+		_ = result.Set(nil)
+		return result
+	}
+
+	var lowerTime, upperTime time.Time
+
+	for i := 0; i < events.Len(); i++ {
+		eventTime := events.At(i).Timestamp().AsTime()
+
+		if lowerTime.IsZero() || eventTime.Before(lowerTime) {
+			lowerTime = eventTime
+		}
+		if upperTime.IsZero() || eventTime.After(upperTime) {
+			upperTime = eventTime
+		}
+	}
+
+	result = pgtype.Tstzrange{
+		Lower:     pgtype.Timestamptz{Time: lowerTime, Status: pgtype.Present},
+		Upper:     pgtype.Timestamptz{Time: upperTime, Status: pgtype.Present},
+		LowerType: pgtype.Inclusive,
+		UpperType: pgtype.Exclusive,
+		Status:    pgtype.Present,
+	}
+
+	return result
+}
+
+func (p *pgxDispatcher) InsertSpan(ctx context.Context, span pdata.Span, nameID, instLibID, rSchemaURLID pgtype.Int8, resourceTags pdata.AttributeMap) error {
+	spanIDInt := convertByteArrayToInt64(span.SpanID().Bytes())
+	parentSpanIDInt := convertByteArrayToInt64(span.ParentSpanID().Bytes())
+	rawResourceTags := resourceTags.AsRaw()
+	if err := p.insertTags(ctx, rawResourceTags, ResourceTagType); err != nil {
+		return err
+	}
+	jsonResourceTags, err := json.Marshal(rawResourceTags)
+	if err != nil {
+		return err
+	}
+	rawTags := span.Attributes().AsRaw()
+	if err := p.insertTags(ctx, rawTags, SpanTagType); err != nil {
+		return err
+	}
+	jsonTags, err := json.Marshal(rawTags)
+	if err != nil {
+		return err
+	}
+
+	eventTimeRange := getEventTimeRange(span.Events())
+
+	_, err = p.conn.Exec(
+		ctx,
+		"INSERT INTO "+schema.Trace+`.span 
+			(trace_id, span_id, trace_state, parent_span_id, name_id, 
+			span_kind, start_time, end_time, span_tags, dropped_tags_count,
+			event_time, dropped_events_count, dropped_link_count, status_code,
+			status_message, instrumentation_lib_id, resource_tags, resource_dropped_tags_count,
+			resource_schema_url_id) 
+			VALUES ($1, $2, $3, $4, $5, $6, $7, $8, `+schema.TracePublic+".get_tag_map($9), $10, $11, $12, $13, $14, $15, $16, "+schema.TracePublic+".get_tag_map($17), $18, $19)",
+		getUUID(span.TraceID().Bytes()),
+		spanIDInt,
+		getTraceStateValue(span.TraceState()),
+		parentSpanIDInt,
+		nameID,
+		span.Kind().String(),
+		span.StartTimestamp().AsTime(),
+		span.EndTimestamp().AsTime(),
+		string(jsonTags),
+		span.DroppedAttributesCount(),
+		eventTimeRange,
+		span.DroppedEventsCount(),
+		span.DroppedLinksCount(),
+		span.Status().Code().String(),
```

Comment:
The code needs to follow the logic https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto#L264 for handling deprecated code logic.

Reply:
Slight issue here, we don't have access to the deprecated code value. `pdata` model doesn't expose it and we don't have direct access to the protobuf struct. 

Only solution I see right now is to clamp the value to 3.